page.go

```diff
@@ -464,6 +464,126 @@ func (p *Page) Screenshot(fullPage bool, req *proto.PageCaptureScreenshot) ([]by
+
+	// Quality (optional) Compression quality from range [0..100] (jpeg only).
+	Quality *int `json:"quality,omitempty"`
+
+	// FixedTop (optional) The number of pixels to skip from the top.
+	// It is suitable for optimizing the screenshot effect when there is a fixed positioning element at the top of the page.
+	FixedTop float64
+
+	// FixedBottom (optional) The number of pixels to skip from the bottom.
+	FixedBottom float64
+
+	// WaitPreScroll wait scroll animation (default is 500ms)
+	WaitPreScroll time.Duration
+}
+
+// ScrollScreenshot Scroll screenshot does not adjust the size of the viewport, but achieves it by scrolling and capturing screenshots in a loop, and then stitching them together.
+// Note that this method also has a flaw: when there are elements with fixed positioning on the page (usually header navigation components), these elements will appear repeatedly,
+//
+//	you can set the FixedTop parameter to optimize it.
+func (p *Page) ScrollScreenshot(opt *ScrollScreenshotOptions) ([]byte, error) {
+	if opt == nil {
+		opt = &ScrollScreenshotOptions{}
+	}
+	if opt.WaitPreScroll == 0 {
+		opt.WaitPreScroll = time.Millisecond * 500
+	}
+
+	metrics, err := proto.PageGetLayoutMetrics{}.Call(p)
+	if err != nil {
+		return nil, err
+	}
+	viewpointHeight := metrics.CSSVisualViewport.ClientHeight
+	contentHeight := metrics.CSSContentSize.Height
+
+	if metrics.CSSContentSize == nil {
+		return nil, errors.New("failed to get css content size")
+	}
+
+	var scrollTop float64
+	var images []utils.ImgWithBox
+
+	for {
+		var clip *proto.PageViewport
+
+		scrollY := viewpointHeight - (opt.FixedTop + opt.FixedBottom)
+		if scrollTop+viewpointHeight > contentHeight {
+			clip = &proto.PageViewport{
+				X:      0,
+				Y:      scrollTop,
+				Width:  metrics.CSSVisualViewport.ClientWidth,
+				Height: contentHeight - scrollTop,
+				Scale:  1,
+			}
+		} else {
+			if scrollTop == 0 {
+				clip = &proto.PageViewport{
+					X:      0,
+					Y:      scrollTop,
+					Width:  metrics.CSSVisualViewport.ClientWidth,
+					Height: scrollY,
+					Scale:  1,
+				}
+			} else {
+				clip = &proto.PageViewport{
+					X:      0,
+					Y:      scrollTop + opt.FixedTop,
+					Width:  metrics.CSSVisualViewport.ClientWidth,
+					Height: scrollY,
+					Scale:  1,
+				}
+			}
+		}
+
+		req := &proto.PageCaptureScreenshot{
+			Format:                opt.Format,
+			Quality:               opt.Quality,
+			Clip:                  clip,
+			FromSurface:           false,
+			CaptureBeyondViewport: false,
+			OptimizeForSpeed:      false,
+		}
+		bs, err := p.Screenshot(false, req)
+		if err != nil {
+			return nil, err
+		}
+
+		images = append(images, utils.ImgWithBox{Img: bs})
+
+		scrollTop += scrollY
+		if scrollTop >= contentHeight {
+			break
+		}
+		err = p.Mouse.Scroll(0, scrollY, 1)
+		if err != nil {
+			return nil, fmt.Errorf("scroll error: %w", err)
+		}
+
+		err = p.WaitStable(opt.WaitPreScroll)
+		if err != nil {
```

Comment:
I can't cover this error branch in my tests, please help, bro.

Reply:
![CleanShot 2024-02-02 at 19 41 23@2x](https://github.com/go-rod/rod/assets/1415488/85f3c972-5ae1-4623-b3c3-af3dc8739c2d)

Reply:
I tried, mentioned in the comments above.

Reply:
Use `g.mc.stubErr(1, proto.DOMGetContentQuads{})`

Reply:
Your code uses `DOMSnapshotCaptureSnapshot`, but we don't want this cdp call to fail right? You want some cdp call in WaitStable to fail, `DOMGetContentQuads` is one that WaitStable uses.

Reply:
> DOMGetContentQuads

It seems like there are still some problems:


```
	g.Panic(func() {
		// mock error for WaitStable
		g.mc.stubErr(1, proto.DOMGetContentQuads{})
		p.MustScrollScreenshotPage()
	})
```

--- FAIL: TestScrollScreenshotPage (47.14s)
    page_test.go:805:  ⦗should panic⦘

Reply:
> Your code uses `DOMSnapshotCaptureSnapshot`, but we don't want this cdp call to fail right? You want some cdp call in WaitStable to fail, `DOMGetContentQuads` is one that WaitStable uses.

Haha, I saw that `WaitDOMStable` would be called `DOMSnapshotCaptureSnapshot` req, so I gave it a try.

Reply:
Got it, let me run the code

Reply:
I pushed a race condition fix on the main branch: https://github.com/go-rod/rod/commit/4d3ff679eaa440737b2dab06f2f4660f83dc6b92

Then this works for me:

```go
	g.Panic(func() {
		// mock error for Screenshot
		g.mc.stubErr(1, proto.DOMSnapshotCaptureSnapshot{})
		p.MustScrollScreenshotPage()
	})
```

Reply:
> I pushed a race condition fix on the main branch: [ce2a1af](https://github.com/go-rod/rod/commit/ce2a1afc29647fd9ca403829333a733fd39bffc9)
> 
> Then this works for me:
> 
> ```go
> 	g.Panic(func() {
> 		// mock error for Screenshot
> 		g.mc.stubErr(1, proto.DOMSnapshotCaptureSnapshot{})
> 		p.MustScrollScreenshotPage()
> 	})
> ```

Yes, I resubmitted the code. Please try running CI again. It’s just a little bit off.